internals/overlord/state/notices.go

```diff
@@ -0,0 +1,386 @@
+			// Update last repeated time if repeat-after time has elapsed
+			notice.lastRepeated = now
+			newOrRepeated = true
+		}
+	}
+	notice.lastOccurred = now
+	notice.lastData = data
+	notice.repeatAfter = repeatAfter
+
+	if newOrRepeated {
+		s.processNoticeWaiters()
+	}
+
+	return notice.id
+}
+
+func uniqueNoticeKey(noticeType NoticeType, key string) string {
+	return string(noticeType) + ":" + key
+}
+
+// NoticeFilters allows filter notices by various fields.
+type NoticeFilters struct {
+	// Type, if set, includes only notices of this type.
+	Type NoticeType
+	// Key, if set, includes only notices with this key.
+	Key string
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// matches reports whether the notice n matches these filters
+func (f NoticeFilters) matches(n *Notice) bool {
+	if f.Type != "" && f.Type != n.noticeType {
+		return false
+	}
+	if f.Key != "" && f.Key != n.key {
+		return false
+	}
+	if !f.After.IsZero() && !n.lastRepeated.After(f.After) {
+		return false
+	}
+	return true
+}
+
+// Notices returns the list of notices that match the filters (if any),
+// ordered by the last-repeated time.
+func (s *State) Notices(filters NoticeFilters) []*Notice {
+	s.reading()
+
+	notices := s.flattenNotices(filters)
+	sort.Slice(notices, func(i, j int) bool {
+		return notices[i].lastRepeated.Before(notices[j].lastRepeated)
+	})
+	return notices
+}
+
+// Notice return a single notice by ID, or nil if not found.
+func (s *State) Notice(id string) *Notice {
+	s.reading()
+
+	// Could use another map for lookup, but the number of notices will likely
+	// be small, and this function is probably only used rarely by the CLI, so
+	// performance is unlikely to matter.
+	for _, notice := range s.notices {
+		if notice.id == id {
+			return notice
+		}
+	}
+	return nil
+}
+
+func (s *State) flattenNotices(filters NoticeFilters) []*Notice {
+	now := time.Now()
+	var notices []*Notice
+	for _, n := range s.notices {
+		if n.expired(now) || !filters.matches(n) {
+			continue
+		}
+		notices = append(notices, n)
+	}
+	return notices
+}
+
+func (s *State) unflattenNotices(flat []*Notice) {
+	now := time.Now()
+	s.notices = make(map[string]*Notice)
+	for _, n := range flat {
+		if n.expired(now) {
+			continue
+		}
+		uniqueKey := uniqueNoticeKey(n.noticeType, n.key)
+		s.notices[uniqueKey] = n
+	}
+}
+
+// WaitNotices waits for notices that match the filters to exist or occur,
+// returning the list of matching notices ordered by the last-repeated time.
+//
+// It waits till there is at least one matching notice or the context times
+// out or is cancelled. If there are existing notices that match the
+// filters, WaitNotices will return them immediately.
+func (s *State) WaitNotices(ctx context.Context, filters NoticeFilters) ([]*Notice, error) {
+	// State is already locked here by the caller, so notices won't be added
```

Comment:
We shouldn't be writing to state here, in `Notices` or in `WaitNotices` -- do you see that we are? We're modifying the `noticeWaiters`, but those are ephemeral waiter data structures, not actually stored state.

Reply:
You are completely right, thanks for explaining.

Reply:
Yeah, we definitely need to protect against concurrent access. But there's a distinction here: state that's persisted to disk in the JSON file, and ephemeral tracking that's not. The `addNoticeWaiter` and `removeNoticeWaiter` methods modify the latter -- the waiter data structures, which aren't (and shouldn't be) persisted to disk, because they only live for the duration of the `WaitNotices` call.

The `s.writing()` method sets `s.modified = true`, which tells the state system that persistent state has been modified and that it should do a snapshot to disk. But that's not the case here: `WaitNotices` doesn't change persisted state (that would be `AddNotice`).

In any case, I can see that the defensive locking check would be good, so I've added a call to `s.reading()` so that we're checking that but not `s.writing()` (to avoid setting `s.modified` which isn't appropriate here):

```go
func (s *State) WaitNotices(ctx context.Context, filter *NoticeFilter) ([]*Notice, error) {
	// This is s.reading() instead of s.writing() because addNoticeWaiter only
	// changes the waiter data structures, not persistent state, so s.modified
	// should not be set.
	s.reading()
```

Hope that makes sense.